modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js

```diff
@@ -735,6 +735,14 @@ const FrontendDataSet = ({
 		});
 	}
 
+	function onItemsChange({itemKey = 'id', itemsChanged}) {
+		const updatedItems = new Map(
+			[...items, ...itemsChanged].map((item) => [item[itemKey], item])
+		);
+
+		setItems(Array.from(updatedItems.values()));
```

Comment:
Very interesting trick, I don't remember seeing this before. The classic solution has quadratic complexity, and this solution makes the conversion in linear complexity. The key is using Map constructor to collapse duplicates into the second value.

```js
new Map([['a', 2], ['a', 1]]) // Map(1) {'a' => 1}
```

Nice! :+1: